### ISSUE TITLE
Roslyn cache, sqlite heavily fragments on NTFS

### ISSUE THREAD

**Issue by tamlin-mike** [dotnet/roslyn#46853]
**Version Used**: Latest VS.

This isn't as much a bug report, as a "heads up".

Can you guess how fast it is to read a 55 MiB file from a spinning rust disk?

Around 500 ms?
Good guess. Try again.

Due to the uncontrolled growth of (and writing to) the Roslyn "Cache", I redirected this to a spinning rust disk.

Today I had a look at fragmentation, and... it's not pretty.

Yeah, an argument could be made for "NTFS is crap, deal with it", but one could also make the argument that "Well written software knows that NTFS is crap, and works around it by preallocating filesize", or defer writing so it can be flushed in larger batches.

To give a taste of how bad it can be - this is on a volume with hundreds of contigous GiB free - here's a sample from the five worst of these "Cache" files. Please realize that this is not a hypothetical synthetic scenario, this is actual real-world data.

Size (MB)|frags|MFT ents|Filename
---------|----:|-------:|--------
11 | 139 |      2 | Xxx\RemoteWorkspace\zzz\sqlite3\storage.ide
14 | 141 |      2 | Xxx\RemoteWorkspace\zzz\sqlite3\storage.ide-wal
15 | 158 |      2 | Xxx\Host\zzz\sqlite3\storage.ide
26 | 351 |      3 | Xxx\RemoteWorkspace\zzz\sqlite3\storage.ide
55 | 622 |      5 | Xxx\Host\zzz\sqlite3\storage.ide

`frags` = # of fragments the file is stored in.
`MFT ents` lists how many NTFS MFT entries this file consumes, due to being so heavily fragmented.
`Xxx` = `User\Local\Microsoft\VisualStudio\Roslyn\Cache`
`zzz` are the various Base-64 encoded project hashes.

So how long would it take to read a 55 MiB file again? Let's say you have an average disk, with around 100MiB/s and 10ms avg seek time. Then some simple math says

500ms for the data
10ms * 622 = 6.2 seconds seek time
= **6.7 seconds**.

And that's in the best possible scenario, if the disk have no other requests to service during this eternity.

Please note this isn't intended as finger-pointing, but rather to provide actual data "from the field" to raise awareness of the issue.

**Comment by CyrusNajmabadi** [comment]
>Due to the uncontrolled growth of (and writing to) the Roslyn "Cache"

Can you define uncontrolled growth?  From above it only sounds like 55mb.  But that's well within bounds of our expected size of this cache.

> Can you guess how fast it is to read a 55 MiB file from a spinning rust disk?

All 55MB should not need to be read in.  It's a DB.  So we'll only be reading in the pages that are needed to support our queries.

> And that's in the best possible scenario, if the disk have no other requests to service during this eternity.

I would def not ever put VS caches on a spinning rust disk.  It's simply not going to end up working well.  It's also well outside the norm that data shows us for where people have their local caches on their machine.  As such, things are designed to work well for the 99.99% case here.  If someone moves the cache like this, then the resultant perf issues would be outside of our design.

> "Well written software knows that NTFS is crap, and works around it by preallocating filesize",

This is sqlite.  I only know of hacky ways to make it preallocate the file size.

> or defer writing so it can be flushed in larger batches.

This is already how it operates.  We do things in memory and then bulk flush that data to disk.

**Comment by tamlin-mike** [comment]
> Can you define uncontrolled growth?

Around a quarter of a gigabyte in the first two weeks previously. Half a gigabyte on last check, about a week after emptying the cache dir.

>  From above it only sounds like 55mb.

Oh, nonono. :-) That's for a single db-file for a single project (I had hoped that was obvious). It's waaay larger in total. Regardless, the report wasn't about the size of the database files, but how badly they become fragmented (on NTFS).

> I would def not ever put VS caches on a spinning rust disk. It's simply not going to end up working well.

So placing them on SSD, where they not only fragment the hell out of the filesystem in the same way, now you end up with the write-amplification wear on the SSD feels better for you? Cool. That's your choice. I have chosen a different path.

> This is sqlite. I only know of hacky ways to make it preallocate the file size.

Since SQLite is now part of the platform (it's in Window SDK), I suggest considering those "hackish workarounds" to fix what the platform team screws up. I mean, there's pretty much zero chance _they_ would ever get off their hands and fix this, whether by finally fixing the totally crap bitmap-allocator in NTFS (that they have left rotting in this state for several **decades**), or hacking the workaround in usermode SQLite code.

>> or defer writing so it can be flushed in larger batches.
>
> This is already how it operates. 

622 fragments for a mere 55MB, giving an average fragment size of 90KB, suggests a batching size of at most a mere 64KB. That would have been OK on FAT-16. For floppies. I hope the display of real-world data makes it clear that batch size a bit one the low side. One or two orders of magnitude low.

**Comment by CyrusNajmabadi** [comment]
> So placing them on SSD, where they not only fragment the hell out of the filesystem in the same way. 

I don't see why this would matter. It's an SSD. 

> now you end up with the write-amplification wear on the SSD 

Numbers here seem like this is a non issue. 

> I suggest considering those "hackish workarounds" 

It's not clear to me that there's an actual issue here that needs addressing. That will have to be established first before any work here is done. 

> Half a gigabyte on last check, about a week after emptying the cache dir.

Can you share out the cache somewhere?  I'd like to look at it. Sounds like some component might be acting up. Thanks!

**Comment by miloush** [comment]
> to raise awareness of the issue

@tamlin-mike can you clarify what the issue is? I see you are unhappy about the number of file fragments, how does it translate into a tangible user experience? Is project loading too slow? Is intellisense slow? Is the disk footprint too large?

**Comment by CyrusNajmabadi** [comment]
note: this is no an accurate way of looking at this system:

> So how long would it take to read a 55 MiB file again? Let's say you have an average disk, with around 100MiB/s and 10ms avg seek time. Then some simple math says

> 500ms for the data
> 10ms * 622 = 6.2 seconds seek time
> = 6.7 seconds.

We don't have operations that need to read the entire DB.  instead, we read from CAS-keyed indices.  These will be in the primary keys of the DB and will almost always hit just a couple of pages tops to either see if there's a hit, and if so what to read for it. Only in the case of CAS-match-hits do we read the larger amounts of data.  i.e. think of it as a Dictionary that different service write/read from, but no one is enumerating in the entirety.

Technically, this could be done just with filepaths + content files.  However, using a DB has virtues of providing ACID semantics so we don' have to worry about issues that arise when we dont' have that.  Also, it performs much better for the sort of double-hop lookups we're doing versus actually using the FS and needing to do multiple IO calls to do these jumps.

--

Anyways, i def look forward to seeing the cache file itself.  I think it will be informative as to what's going on here.

**Comment by CyrusNajmabadi** [comment]
FYI, each `zzz` refers to a different *solution*.  So if you have a lot of solutoins, this will build up.  I'm curious not about the total size here, but what max size you're seeing per-solutoin.

Also, one thing we def could do to improve here is to reduce the amount of storage duplication we have.  one core issue is that we make one per roslyn process/workspace we have.  This is because even though the cache would be shareable for most features in all processes/workspaces, there is one critical feature for which this is not true.  Specifically, the Diagnostics subsystem expects to be able to read/write without the data changing by other processes.

@mavasani If we could make it so that the diagnostics sysstem was somehow resilient to this (potentially by using content-addressing to keep different processes from stomping on each other, or just having a single process for persisting and reading this info), we could move back to a shared db here, which would potentially halve/third our usage.

With your work to move diagnostics OOP, do you think we might be able to do that?

**Comment by CyrusNajmabadi** [comment]
Note: i just  checked my own roslyn cache (wihch has been used for months).  It's at 75MB for Roslyn.sln (a large solution).  This is in line with my expectation.  So if you're seeing a gig used for a single DB, i def want to look at that to see what's going on with it.

**Comment by sharwell** [comment]
I'm tentatively marking this as **Won't Fix** for three reasons:

1. The cache is operating within the expected size bounds, and well within the available space of entry-level storage solutions
2. The SQLite implementation is external to this project, and our understanding on multiple reviews is we are using it as it was intended, so either the fragmentation is unavoidable at this point or the fragmentation is not known to directly produce performance problems
3. The performance disadvantage of spinning disks is well-established. A user relying on one at this point would already know that it would produce unavoidably sub-optimal performance, with well-known mitigations that are outside the control of this project

**Comment by CyrusNajmabadi** [comment]
I am ok with that resolution.  If we get more data from teh customer (including a copy of their worst DBs) we can take a look to see if anything is out of line.  

Note: one downside of our current approach is that we make our DBs forward/back compatible.  specifically, when we change the format, we don't touch the old db and we generate a new db that sits next to the old one.  this means that the old DB is still available for the user if they are running multiple versions of VS side by side.  But we never really know if there's a point at which we could remove any of the old DBs.  

However, user can always just torch the directory.  It's completely designed to be optional.

**Comment by miloush** [comment]
I am on a normal HDD (I assume that's what spinning rust disk refers to). The performance of VS is nothing what it used to be, but so are the capabilities. Yes it is sub-optimal, but usable (although I do feel it has regressed recently, and I saw "Please wait for the editor command to complete" box couple of times while simply typing). Anyway, all the serious performance drops that I occasionally run into during writing code are connected to high CPU usage, I never suspected the HDD access to be the bottleneck. Yes the .NET Core performance is way worse than FW probably because the number of files it is shamelessly using, and the time from F5 to a running app is really bad, but I don't think that goes after Roslyn, definitely not the cache. That's why I was enquiring on what practical manifestation does the OP see connected to the fragmentation, since for me, another spinning user, the fragmented 50MB files are a non-issue.

**Comment by CyrusNajmabadi** [comment]
Closing out absent additional user data.  If you can answer some of hte above questions/requests @tamlin-mike that would be appreciated :)
